src/ffs/ffs.ts

```diff
@@ -0,0 +1,104 @@
+import pubTree, { PublicTree } from './pub/tree'
+import privTree, { PrivateTree } from './priv/tree'
+import pubNode, { PublicNode } from './pub/node'
+import { Tree, Node, Link } from './types'
+import { CID, FileContent } from '../ipfs'
+import pathUtil from './path'
+
+export class FFS {
+
+  root: PublicNode
+  pubTree: PublicTree
+  privTree: PrivateTree
+
+  constructor(root: PublicNode, pubTree: PublicTree, privTree: PrivateTree) {
+    this.root = root
+    this.pubTree = pubTree
+    this.privTree = privTree
+  }
+
+  async put(): Promise<CID> {
+    return this.root.put()
+  }
+
+  async cid(): Promise<CID> {
+    return this.put()
+  }
+
+  async listDir(path: string): Promise<Link[] | null> {
+    const { tree, tail } = this.parsePath(path)
+    return tree.listDir(tail)
+  }
+
+  async makeDir(path: string): Promise<FFS> {
+    const { tree, tail } = this.parsePath(path)
+    await tree.makeDir(tail)
+    return this.updateRoot()
+  }
+
+  async addFile(path: string, content: FileContent): Promise<FFS> {
+    const { tree, tail } = this.parsePath(path)
```

Comment:
`tail`?

Reply:
yup its a file path. I could use `relPath` or something similar instead

Reply:
Yeah, that would be clearer 👍